File: frontend/src/api/globalIntelligence.ts
/** API client and types for Global Payment Intelligence. */

import { apiClient } from './client';

export interface GlobalKPIs {
  total_volume: number;
  total_transactions: number;
  success_rate: number;
  failure_rate: number;
  recovery_rate: number;
  revenue_at_risk: number;
  recovered_revenue: number;
  currency_symbol: string;
}

export interface GlobalHealthScore {
  overall_score: number;
  status_label: 'HEALTHY' | 'WATCH' | 'DEGRADED';
  authorization_score: number;
  recovery_score: number;
  gateway_stability_score: number;
  customer_friction_score: number;
  is_demo_derived: boolean;
}

export interface RegionPerformanceItem {
  region_id: string;
  region_name: string;
  country_code: string;
  flag_emoji: string;
  currency: string;
  total_volume: number;
  success_rate: number;
  failure_rate: number;
  recovery_rate: number;
  revenue_at_risk: number;
  recovered_revenue: number;
  status: 'HEALTHY' | 'WATCH' | 'HIGH RISK';
  top_failure_type: string;
  top_gateway: string;
  transaction_count: number;
  coordinates: {
    lat?: number;
    lng?: number;
    x: number;
    y: number;
  };
}

export interface PaymentMethodPerformanceItem {
  method_id: string;
  method_label: string;
  success_rate: number;
  failure_rate: number;
  recovery_rate: number;
  total_volume: number;
  revenue_at_risk: number;
  transaction_count: number;
  is_best_performing: boolean;
  is_highest_risk: boolean;
}

export interface GatewayGlobalPerformanceItem {
  gateway_name: string;
  authorization_rate: number;
  failure_rate: number;
  timeout_rate: number;
  revenue_impact: number;
  status: string;
}

export interface FailureIntelligenceItem {
  failure_type: string;
  failure_label: string;
  count: number;
  volume: number;
  revenue_at_risk: number;
  percentage_of_total: number;
  recovery_rate: number;
}

export interface HeatmapCell {
  region: string;
  failure_type: string;
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  amount: number;
  count: number;
}

export interface GlobalPaymentFunnelStep {
  step_key: string;
  step_label: string;
  count: number;
  percentage: number;
  volume: number;
}

export interface RecoveryOpportunityHighlight {
  recoverable_revenue: number;
  largest_opportunity_category: string;
  largest_opportunity_amount: number;
  best_next_action: string;
  expected_yield_lift: number;
}

export interface TopLeakageAreaItem {
  rank: number;
  failure_type: string;
  failure_label: string;
  revenue_at_risk: number;
  recovery_potential: number;
  percentage: number;
}

export interface GlobalIntelligenceResponse {
  kpis: GlobalKPIs;
  health_score: GlobalHealthScore;
  regions: RegionPerformanceItem[];
  payment_methods: PaymentMethodPerformanceItem[];
  gateways: GatewayGlobalPerformanceItem[];
  failure_intelligence: FailureIntelligenceItem[];
  heatmap: HeatmapCell[];
  funnel: GlobalPaymentFunnelStep[];
  recovery_opportunity: RecoveryOpportunityHighlight;
  top_leakage_areas: TopLeakageAreaItem[];
  executive_summary: string;
  insights: string[];
  technical_signals: Record<string, string>;
  last_updated: string;
  is_simulation: boolean;
}

export interface GlobalIntelligenceFilterParams {
  region?: string;
  gateway?: string;
  payment_method?: string;
  failure_type?: string;
}

export const getGlobalPaymentIntelligence = async (
  params?: GlobalIntelligenceFilterParams
): Promise<GlobalIntelligenceResponse> => {
  const res = await apiClient.get<GlobalIntelligenceResponse>('/global-intelligence/summary', {
    params,
  });
  return res.data;
};

File: backend/app/schemas/global_intelligence.py
"""Pydantic schemas for Global Payment Intelligence."""

from datetime import datetime
from decimal import Decimal
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field


class GlobalKPIs(BaseModel):
    total_volume: Decimal
    total_transactions: int
    success_rate: float
    failure_rate: float
    recovery_rate: float
    revenue_at_risk: Decimal
    recovered_revenue: Decimal
    currency_symbol: str = "₹"


class GlobalHealthScore(BaseModel):
    overall_score: int
    status_label: str  # HEALTHY, WATCH, DEGRADED
    authorization_score: int
    recovery_score: int
    gateway_stability_score: int
    customer_friction_score: int
    is_demo_derived: bool = True


class RegionPerformanceItem(BaseModel):
    region_id: str
    region_name: str
    country_code: str
    flag_emoji: str
    currency: str
    total_volume: Decimal
    success_rate: float
    failure_rate: float
    recovery_rate: float
    revenue_at_risk: Decimal
    recovered_revenue: Decimal
    status: str  # HEALTHY, WATCH, HIGH RISK
    top_failure_type: str
    top_gateway: str
    transaction_count: int
    coordinates: Dict[str, float] = Field(default_factory=dict)


class PaymentMethodPerformanceItem(BaseModel):
    method_id: str
    method_label: str
    success_rate: float
    failure_rate: float
    recovery_rate: float
    total_volume: Decimal
    revenue_at_risk: Decimal
    transaction_count: int
    is_best_performing: bool = False
    is_highest_risk: bool = False


class GatewayGlobalPerformanceItem(BaseModel):
    gateway_name: str
    authorization_rate: float
    failure_rate: float
    timeout_rate: float
    revenue_impact: Decimal
    status: str  # OPTIMAL, HEALTHY, DEGRADED


class FailureIntelligenceItem(BaseModel):
    failure_type: str
    failure_label: str
    count: int
    volume: Decimal
    revenue_at_risk: Decimal
    percentage_of_total: float
    recovery_rate: float


class HeatmapCell(BaseModel):
    region: str
    failure_type: str
    risk_level: str  # LOW, MEDIUM, HIGH, CRITICAL
    amount: Decimal
    count: int


class GlobalPaymentFunnelStep(BaseModel):
    step_key: str
    step_label: str
    count: int
    percentage: float
    volume: Decimal


class RecoveryOpportunityHighlight(BaseModel):
    recoverable_revenue: Decimal
    largest_opportunity_category: str
    largest_opportunity_amount: Decimal
    best_next_action: str
    expected_yield_lift: float


class TopLeakageAreaItem(BaseModel):
    rank: int
    failure_type: str
    failure_label: str
    revenue_at_risk: Decimal
    recovery_potential: Decimal
    percentage: float


class GlobalIntelligenceResponse(BaseModel):
    kpis: GlobalKPIs
    health_score: GlobalHealthScore
    regions: List[RegionPerformanceItem]
    payment_methods: List[PaymentMethodPerformanceItem]
    gateways: List[GatewayGlobalPerformanceItem]
    failure_intelligence: List[FailureIntelligenceItem]
    heatmap: List[HeatmapCell]
    funnel: List[GlobalPaymentFunnelStep]
    recovery_opportunity: RecoveryOpportunityHighlight
    top_leakage_areas: List[TopLeakageAreaItem]
    executive_summary: str
    insights: List[str]
    technical_signals: Dict[str, str]
    last_updated: datetime
    is_simulation: bool = True
